### ISSUE TITLE
Issue with custom Regex Library

### ISSUE THREAD

**Issue by schneefisch** [santhosh-tekuri/jsonschema#148]
Hello,

In recent Changes (see issue https://github.com/santhosh-tekuri/jsonschema/pull/61) and commit (https://github.com/santhosh-tekuri/jsonschema/commit/bbc9f8974f61b76d148d1230709a653234e619bd) you have added the possibility to use a custom regex-parser.

E.g. by importing `github.com/dlclark/regexp2`:

```go
compiler := jsonschema.NewCompiler()
compiler.Draft = jsonschema.Draft2020 
compiler.CompileRegex = func(s string) (jsonschema.Regexp, error) {
	re, _ := regexp2.Compile(s, regexp2.ECMAScript)
	return (*customGoRegexp)(re), nil
}
compiler.Formats = jsonschema.Formats
compiler.Formats["regex"] = customIsRegex

newJsonSchema, err := compiler.Compile(schemaPath)
```
Also, we have changed the `isRegex` inside of Formats

Sadly, there are (at least two) issues with that.

At first, our Custom `isRegex` is not called at the `compiler.Compile(schemaPath)` stage, instead the default one is used.
This is likely related to a direct call to `isRegex` in the schema.go (https://github.com/santhosh-tekuri/jsonschema/blob/cee473f3136c6dd6f2507e55182c671af6e6f160/schema.go#L352), this could be replaced with

```go
if s.RegexProperties {
	for pname := range v {
		if !s.format(pname) {
			errors = append(errors, validationError("", "patternProperty %s is not valid regex", quote(pname)))
		}
	}
}
```

Also there seems to be a subsequent issue, even if I am doing that, then the s.format does not return my custom configured `customIsRegex` function but instead still has the default one configured. So there might be a subsequent issue that prevents from configuring the `jsonschema.Formats["regex"]`.
The exact call is in schema.go line 305 (https://github.com/santhosh-tekuri/jsonschema/blob/cee473f3136c6dd6f2507e55182c671af6e6f160/schema.go#L305). This is called during compilation of the schema, but here the overloaded format validator is not used. It might be, that it is in in compiler.go line 795 (https://github.com/santhosh-tekuri/jsonschema/blob/cee473f3136c6dd6f2507e55182c671af6e6f160/compiler.go#L791) where the `compiler.Formats` are not passed on to the validation.

Secondly, in compiler.go line 573 (https://github.com/santhosh-tekuri/jsonschema/blob/cee473f3136c6dd6f2507e55182c671af6e6f160/compiler.go#L573) is also a direct call to the go-native regex, which is not replaced by the custom configured regex library.
This place must also be changed in order to enable setting a different, custom regex library, here I don't know yet how to do that.

A sample-pattern I am using to test this is `^(?!delta)(.{1,27})$`

long story short: It looks like the RegexCompiler which is configured on jsonschema.Compiler is only used when checking a dataset against a schema. During the validation as part of the compilation of the schema itself, the custom RegexCompiler is not used for validating the regexes mentioned in fields with `format:regex` or the keys of `patternProperties`-Objects.

**Comment by santhosh-tekuri** [comment]
@schneefisch 
as you pointed out I am still using native-regex. I can fix those things that you mentioned.

but there is another issue. the draft metaschema are precompiled in package [init](https://github.com/santhosh-tekuri/jsonschema/blob/cee473f3136c6dd6f2507e55182c671af6e6f160/draft.go#L256)
this pre-compilation will end up using go-native regex.

in current implementation we precompile draft meta schemas once during package initialisation and use it during [user schema compilation](https://github.com/santhosh-tekuri/jsonschema/blob/cee473f3136c6dd6f2507e55182c671af6e6f160/compiler.go#L786)

so this feature becomes half backed.

one choice is compile draft metaschemas for each user schema compilation which seems redundant.

do you have any suggestions.

**Comment by schneefisch** [comment]
Thats an interesting suggestion, I did not see that so far.
so sadly I do not yet have a suggestion. Maybe I can find something.

**Comment by santhosh-tekuri** [comment]
I think we can recompile draft metaschemas only when user has overridden regex impl in compiler.

This should not affect the performance for those who are using native-regex.

let me work on this

**Comment by santhosh-tekuri** [comment]
@schneefisch  

I have fixed main issues you have mentioned with d33a6df

I have tested it with following program:
```go
package main

import (
        "encoding/json"
        "strings"

        "github.com/dlclark/regexp2"
        "github.com/santhosh-tekuri/jsonschema/v5"
)

var schema = `{
        "patternProperties": {
                "^a.*" : { "type": "number" }
        }
}`
var instance = `{ "a": "34"}`

func main() {
        compiler := jsonschema.NewCompiler()
        compiler.CompileRegex = func(s string) (jsonschema.Regexp, error) {
                println("compiling regex: ", s)
                re, err := regexp2.Compile(s, regexp2.ECMAScript)
                if err != nil {
                        return nil, err
                }
                return ecmaRegex{re}, nil
        }
        compiler.Formats["regex"] = func(v interface{}) bool {
                println("checking regex format")
                s, ok := v.(string)
                if !ok {
                        return true
                }
                _, err := regexp2.Compile(s, regexp2.ECMAScript)
                return err == nil
        }

        if err := compiler.AddResource("schema.json", strings.NewReader(schema)); err != nil {
                panic(err)
        }
        sch, err := compiler.Compile("schema.json")
        if err != nil {
                panic(err)
        }

        var inst interface{}
        if err := json.Unmarshal([]byte(instance), &inst); err != nil {
                panic(err)
        }
        if err := sch.Validate(inst); err != nil {
                panic(err)
        }
}

type ecmaRegex struct {
        re *regexp2.Regexp
}

func (re ecmaRegex) MatchString(s string) bool {
        println("matching regex: ", s)
        matched, err := re.re.MatchString(s)
        return err != nil && matched
}

func (re ecmaRegex) String() string {
        return re.re.String()
}
```

when run it does prints from custom regex implementation as below:
```
$ go run .
compiling regex:  ^a.*
matching regex:  a
```

still the draft meta-schemas issue is not yet fixed. so there is still following limitation:

consider a schema as below:
```json
{
    "patternProperties": {
        "XXX": {}
    }
}
```

if `XXX` is valid regex when using `github.com/dlclark/regexp2` but not with native-regex, you would expect the above schema to compile. but it fails to compile because the schema itself is validated with draft meta-schema(which is currently compiled with native-regex)

**Comment by santhosh-tekuri** [comment]
@schneefisch 

now meta schemas are compiled with custom regex engine if needed. see 333d938

this should fix the issue completely. could you please re-verify at your end.

**Comment by schneefisch** [comment]
Wow, that was quick
I will check that out immediately ;)

**Comment by schneefisch** [comment]
Indeed, My use-case seems fixed. 
Thank you so much for that lightning-speed response and fix. That was amazing.

**Comment by santhosh-tekuri** [comment]
637fbb2 with this commit you no longer need to register `Compiler.Foramts["regex"]`

**Comment by santhosh-tekuri** [comment]
d265bc9 adds example test case
